Offline/OfflinePlayer/objc/OfflinePlayer/DownloadsViewController.m

```diff
@@ -0,0 +1,1181 @@
+	{
+		// iOS 11.2+: cancel normally
+		[self cancelVideoDownload];
+	}
+	else
+	{
+		if (@available(iOS 11.0, *))
+		{
+			// iOS 11.0 and 11.1: work around iOS download manager bugs
+			[self forceStopAllDownloadTasks];
+		}
+		else
+		{
+			// iOS 10.x: cancel normally
+			[self cancelVideoDownload];
+		}
+	}
+}
+
+- (void)forceStopAllDownloadTasks
+{
+	// iOS 11.0 and 11.1 have a bug in which some downloads cannot be stopped using normal methods.
+	// As a workaround, you can call "forceStopAllDownloadTasks" to cancel all the video downloads
+	// that are still in progress.
+	UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Stop All Downloads"
+																   message:@"Do you want to stop all the downloads in progress?"
+															preferredStyle:UIAlertControllerStyleAlert];
+	UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:@"Stop All" style:UIAlertActionStyleDefault
+														  handler:^(UIAlertAction * action) {
+
+															  [BCOVOfflineVideoManager.sharedManager forceStopAllDownloadTasks];
+
+														  }];
+	UIAlertAction* cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel
+														 handler:^(UIAlertAction * action) {}];
+
+	[alert addAction:defaultAction];
+	[alert addAction:cancelAction];
+	[self presentViewController:alert animated:YES completion:nil];
+}
+
+- (void)cancelVideoDownload
+{
+	BCOVOfflineVideoManager *sharedManager = BCOVOfflineVideoManager.sharedManager;
+
+	BCOVOfflineVideoStatus *offlineVideoStatus = [sharedManager offlineVideoStatusForToken:self.selectedOfflineVideoToken];
+
+	switch (offlineVideoStatus.downloadState)
+	{
+		case BCOVOfflineVideoDownloadStateRequested:
+		case BCOVOfflineVideoDownloadStateTracksRequested:
+		case BCOVOfflineVideoDownloadStateDownloading:
+		case BCOVOfflineVideoDownloadStateTracksDownloading:
+		case BCOVOfflineVideoDownloadStateSuspended:
+		case BCOVOfflineVideoDownloadStateTracksSuspended:
+			[sharedManager cancelVideoDownload:self.selectedOfflineVideoToken];
+			break;
+
+		case BCOVOfflineVideoDownloadLicensePreloaded:
+		case BCOVOfflineVideoDownloadStateCancelled:
+		case BCOVOfflineVideoDownloadStateTracksCancelled:
+		case BCOVOfflineVideoDownloadStateCompleted:
+		case BCOVOfflineVideoDownloadStateTracksCompleted:
+		case BCOVOfflineVideoDownloadStateError:
+		case BCOVOfflineVideoDownloadStateTracksError:
+			break;
+	}
+}
+
+- (void)viewDidLoad
+{
+    [super viewDidLoad];
+
+    gDownloadsViewController = self;
+
+    self.tabBarController = (UITabBarController*)self.parentViewController;
+
+    self.downloadsTableView.dataSource = self;
+    self.downloadsTableView.delegate = self;
+    [self.downloadsTableView setContentInset:UIEdgeInsetsMake(0, 0, 0, 0)];
+    [self createTableFooter];
+    [self updateInfoForSelectedDownload];
+    
+    [self createPlayerView];
+
+    {
+        UILongPressGestureRecognizer *longPress = [[UILongPressGestureRecognizer alloc]
```

Comment:
Insert comment stolen from below:
"// Long press on a downloaded video gives the option of downloading all tracks"

Reply:
√